policy/compressor.go

```diff
@@ -0,0 +1,132 @@
+package policy
+
+import (
+	"sync"
+
+	"github.com/m3db/m3cluster/kv"
+	"github.com/m3db/m3metrics/generated/proto/schema"
+	"github.com/m3db/m3x/instrument"
+	"github.com/m3db/m3x/log"
+)
+
+// Compressor maintains a mapping of policies to integers. It is useful when encoding policies
+// so that an integer representation of a policy can be sent instead of the entire Policy
+type Compressor interface {
+	Get(p Policy) (int64, bool)
+}
+
+type compressorMap map[Policy]int64
+
+type noopCompressor struct{}
+
+// NewNoopCompressor returns a new Compressor which never matches a policy
+func NewNoopCompressor() Compressor {
+	return noopCompressor{}
+}
+
+func (m noopCompressor) Get(p Policy) (int64, bool) {
+	return -1, false
+}
+
+type staticCompressor struct {
+	policies compressorMap
+}
+
+// NewStaticCompressor returns a new static Compressor
+func NewStaticCompressor(policies map[Policy]int64) Compressor {
+	return staticCompressor{policies: policies}
+}
+
+func (m staticCompressor) Get(p Policy) (int64, bool) {
+	i, ok := m.policies[p]
+	return i, ok
+}
+
+type dynamicCompressor struct {
+	sync.RWMutex
+
+	policies compressorMap
+	log      xlog.Logger
+}
+
+// NewDynamicCompressor returns a new dynamic Compressor
+func NewDynamicCompressor(w kv.ValueWatch, opts instrument.Options) Compressor {
+	m := &dynamicCompressor{
+		policies: make(compressorMap),
+		log:      opts.Logger(),
+	}
+	m.update(w.Get())
+
+	go m.run(w)
```

Comment:
That's cool with me. I don't think we'd leak any goroutines in this case though since the goroutine will return when the `ValueWatch` is closed.